File: src/data-collector.js
const fs = require('fs-extra');
const path = require('path');
const config = require('./config');

class DataCollector {
  constructor() {
    this.results = [];
    this.filename = this.generateFilename();
    this.ensureResultsDir();
  }

  generateFilename() {
    const now = new Date();
    const timestamp = now.toISOString().replace(/[:.]/g, '-').slice(0, 19);
    return `experiment_${timestamp}.csv`;
  }

  async ensureResultsDir() {
    await fs.ensureDir(config.output.resultsDir);
  }

  addResult(apiType, queryType, concurrentClients, cacheState, result) {
    const record = {
      timestamp: new Date().toISOString(),
      api_type: apiType,
      query_type: queryType,
      concurrent_clients: concurrentClients,
      cache_state: cacheState,
      response_time_ms: result.responseTime,
      payload_size_bytes: result.payloadSize,
      status_code: result.statusCode
    };
    
    this.results.push(record);
  }

  async saveResults() {
    const filepath = path.join(config.output.resultsDir, this.filename);
    
    // Create CSV content
    const headers = config.output.csvHeaders.join(',');
    const rows = this.results.map(record => 
      config.output.csvHeaders.map(header => record[header]).join(',')
    );
    
    const csvContent = [headers, ...rows].join('\n');
    
    await fs.writeFile(filepath, csvContent);
    console.log(`Results saved to: ${filepath}`);
    return filepath;
  }

  getStats() {
    return {
      totalMeasurements: this.results.length,
      successfulRequests: this.results.filter(r => r.status_code === 200).length,
      failedRequests: this.results.filter(r => r.status_code !== 200).length,
      avgResponseTime: this.results.reduce((sum, r) => sum + r.response_time_ms, 0) / this.results.length,
      avgPayloadSize: this.results.reduce((sum, r) => sum + r.payload_size_bytes, 0) / this.results.length
    };
  }
}

module.exports = DataCollector;

File: src/data_collector.py
import csv
import os
from datetime import datetime
from config import Config

class DataCollector:
    def __init__(self):
        self.results = []
        self.filename = self._generate_filename()
        self._ensure_results_dir()
    
    def _generate_filename(self):
        timestamp = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
        return f'experiment_{timestamp}.csv'
    
    def _ensure_results_dir(self):
        os.makedirs(Config.RESULTS_DIR, exist_ok=True)
    
    def add_result(self, api_type, query_type, concurrent_clients, cache_state, result):
        record = {
            'timestamp': datetime.now().isoformat(),
            'api_type': api_type,
            'query_type': query_type,
            'concurrent_clients': concurrent_clients,
            'cache_state': cache_state,
            'response_time_ms': result['response_time'],
            'payload_size_bytes': result['payload_size'],
            'status_code': result['status_code']
        }
        self.results.append(record)
    
    def save_results(self):
        filepath = os.path.join(Config.RESULTS_DIR, self.filename)
        
        with open(filepath, 'w', newline='', encoding='utf-8') as csvfile:
            writer = csv.DictWriter(csvfile, fieldnames=Config.CSV_HEADERS)
            writer.writeheader()
            writer.writerows(self.results)
        
        print(f"Results saved to: {filepath}")
        return filepath
    
    def get_stats(self):
        if not self.results:
            return {}
        
        successful = [r for r in self.results if r['status_code'] == 200]
        failed = [r for r in self.results if r['status_code'] != 200]
        
        avg_response_time = sum(r['response_time_ms'] for r in self.results) / len(self.results)
        avg_payload_size = sum(r['payload_size_bytes'] for r in self.results) / len(self.results)
        
        return {
            'total_measurements': len(self.results),
            'successful_requests': len(successful),
            'failed_requests': len(failed),
            'success_rate': len(successful) / len(self.results) * 100,
            'avg_response_time_ms': round(avg_response_time, 2),
            'avg_payload_size_bytes': round(avg_payload_size, 2)
        }